src/Interpreters/Aggregator.cpp

```diff
@@ -26,6 +27,133 @@
 #include <Interpreters/JIT/CompiledExpressionCache.h>
 #include <Core/ProtocolDefines.h>
 
+#include <Parsers/ASTSelectQuery.h>
+
+namespace
+{
+/** Collects observed HashMap-s sizes to avoid redundant intermediate resizes.
+  */
+class HashTablesStatistics
+{
+public:
+    using Params = DB::Aggregator::Params::StatsCollectingParams;
+
+    std::optional<size_t> getSizeHint(const Params & params)
+    {
+        std::lock_guard lock(mutex);
+        const auto cache = getHashTableStatsCache(params, lock);
+        if (const auto hint = cache->get(params.key))
+        {
+            LOG_DEBUG(&Poco::Logger::get("Aggregator"), "An entry for key={} found in cache: size_hint={}", params.key, *hint);
+            return *hint;
+        }
+        return std::nullopt;
+    }
+
+    void update(size_t observed_size, const Params & params)
+    {
+        std::lock_guard lock(mutex);
+        const auto cache = getHashTableStatsCache(params, lock);
+        const auto hint = cache->get(params.key);
+        // We'll maintain the maximum among all the observed values until the next prediction turns out to be too wrong.
+        if (!hint || observed_size > *hint || observed_size < *hint / 2)
+        {
+            LOG_DEBUG(&Poco::Logger::get("Aggregator"), "Statictics updated for key={}: new size_hint={}", params.key, observed_size);
+            cache->set(params.key, std::make_shared<size_t>(observed_size));
+        }
+    }
+
+private:
+    using Cache = DB::LRUCache<UInt64, size_t>;
+    using CachePtr = std::shared_ptr<Cache>;
+
+    CachePtr getHashTableStatsCache(const Params & params, [[maybe_unused]] std::lock_guard<std::mutex> & cache_lock)
+    {
+        if (!hash_table_stats || hash_table_stats->maxSize() != params.max_entries_for_hash_table_stats)
+            hash_table_stats = std::make_shared<Cache>(params.max_entries_for_hash_table_stats);
+        return hash_table_stats;
+    }
+
+    std::mutex mutex;
+    CachePtr hash_table_stats;
```

Comment:
Worth exporting this LRUCache in the same manner as others, in AsynchronousMetrics.cpp:

https://github.com/ClickHouse/ClickHouse/blob/b1d2bdf5699c3726dc7f6fd3bca744eb267c187a/src/Interpreters/AsynchronousMetrics.cpp#L545-L551

Reply:
thx, I replaced in the description the part about `SYSTEM DROP HASH TABLE STATISTICS` with what you suggested.

Reply:
done